src/protocols/secure_channel/CASESession.cpp

```diff
@@ -1119,79 +1012,108 @@ void CASESession::SendErrorMsg(SigmaErrorType errorCode)
+
+    Encoding::LittleEndian::BufferWriter bbuf(destinationMessage, sizeof(destinationMessage));
 
-    trustedRoot = CertificateKeyId(trustedRootId);
+    bbuf.Put(random.data(), random.size());
+    bbuf.Put(rootPubkey.data(), rootPubkey.size());
+    bbuf.Put64(fabricId);
+    bbuf.Put64(nodeId);
+
+    VerifyOrReturnError(bbuf.Fit(), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    ReturnErrorOnFailure(hmac.HMAC_SHA256(ipk.data(), ipk.size(), destinationMessage, sizeof(destinationMessage),
+                                          destinationId.data(), destinationId.size()));
+
+    return CHIP_NO_ERROR;
+}
 
-    for (uint32_t i = 0; i < nTrustedRoots; ++i)
+CHIP_ERROR CASESession::FindDestinationIdCandidate(const ByteSpan & destinationId, const ByteSpan & initiatorRandom,
+                                                   const ByteSpan * ipkList, size_t ipkListEntries)
+{
+    uint8_t nCertificateSets = mOpCredSet->GetCertCount();
+
+    for (uint8_t i = 0; i < nCertificateSets; ++i)
     {
-        ReturnErrorOnFailure(tlvReader.FindElementWithTag(CASETLVTag::kTrustedRootID, suppTlvReader));
+        uint8_t candidate[kSHA256_Hash_Length] = { 0 };
+        CertificateKeyId trustedRootId;
+        ChipCertificateData nodeOperationalCertificate;
+        NodeId nodeId;
+        FabricId fabricId;
+
+        trustedRootId = mOpCredSet->GetTrustedRootId(i);
 
-        VerifyOrReturnError(kTrustedRootIdSize == suppTlvReader.GetLength(), CHIP_ERROR_INVALID_TLV_ELEMENT);
-        VerifyOrReturnError(suppTlvReader.GetType() == TLV::kTLVType_ByteString, CHIP_ERROR_WRONG_TLV_TYPE);
-        ReturnErrorOnFailure(suppTlvReader.GetBytes(trustedRootId, kTrustedRootIdSize));
+        ReturnErrorOnFailure(DecodeChipCert(mOpCredSet->GetDevOpCred(trustedRootId), mOpCredSet->GetDevOpCredLen(trustedRootId),
+                                            nodeOperationalCertificate));
 
-        if (mOpCredSet->IsTrustedRootIn(trustedRoot))
+        ReturnErrorOnFailure(nodeOperationalCertificate.mSubjectDN.GetCertChipId(nodeId));
+        ReturnErrorOnFailure(nodeOperationalCertificate.mSubjectDN.GetCertFabricId(fabricId));
+
+        const ChipCertificateData * rootCertificate = mOpCredSet->GetRootCertificate(trustedRootId);
+        VerifyOrReturnError(rootCertificate != nullptr, CHIP_ERROR_CERT_NOT_FOUND);
+        VerifyOrReturnError(!rootCertificate->mPublicKey.empty(), CHIP_ERROR_INTERNAL);
+        VerifyOrReturnError(rootCertificate->mPublicKey.size() == kP256_PublicKey_Length, CHIP_ERROR_INTERNAL);
+
+        for (size_t j = 0; j < ipkListEntries; ++j)
         {
-            if (!mTrustedRootId.empty())
+            MutableByteSpan candidateSpan(candidate, sizeof(candidate));
+            ReturnErrorOnFailure(
+                GenerateDestinationID(initiatorRandom, rootCertificate->mPublicKey, nodeId, fabricId, ipkList[j], candidateSpan));
+
+            if (memcmp(destinationId.data(), candidate, destinationId.size()) == 0)
```

Comment:
You can use:
`if (destinationId.data_equal(candidate))` for these memcmp checks

Reply:
Updated.

Reply:
Updated.